reconcile/skupper_network/reconciler.py

```diff
@@ -0,0 +1,269 @@
+
+    Connection sites together algorithm:
+    1. Check if all related sites (namespaces) are available
+    2. Iterate over all "to be connected sites" (connected_sites) and
+    2.1. if we are already connected (connection-token exists) -> done
+    2.2. if not connected (connection-token doesn't exist) -> token exchange alghorithm
+
+    Token exchange alghorithm:
+    1. Create a connection-token-request (secret with a unique token name) in the connected site's namespace
+    2. Wait for the site controller to process the connection-token-request
+    3. Move the connection-token (the processed connection-token-request secret) from the connected site's namespace into the site's namespace
+    """
+    oc = oc_map.get_cluster(site.cluster.name)
+    if not oc.project_exists(site.namespace.name):
+        logging.info(
+            f"{site}: Namespace does not exist (yet). Skipping this skupper site for now!"
+        )
+        return
+
+    for connected_site in site.connected_sites:
+        # an existing connection token means we are already connected to this site
+        if not _get_token(oc_map, site, site.token_name(connected_site)):
+            # no connection token found. get the token from the connected site and import it
+            connected_site_oc = oc_map.get_cluster(connected_site.cluster.name)
+            if not connected_site_oc.project_exists(connected_site.namespace.name):
+                logging.info(
+                    f"{connected_site}: Namespace does not exist (yet)."
+                    " Skipping this skupper site for now!"
+                )
+                continue
+
+            if token := _get_token(
+                oc_map,
+                connected_site,
+                connected_site.unique_token_name(site),
+            ):
+                # token found - move it to this site
+                _transfer_token(
+                    oc_map,
+                    site,
+                    connected_site,
+                    dry_run,
+                    integration,
+                    integration_version,
+                    token,
+                )
+            else:
+                # no token found - create a new connection token to be used by this site
+                _create_token(
+                    oc_map,
+                    site,
+                    connected_site,
+                    dry_run,
+                    integration,
+                    integration_version,
+                )
+
+
+def delete_unused_tokens(site: SkupperSite, oc_map: OC_Map, dry_run: bool) -> None:
+    """Delete any other connection tokens that are no longer needed."""
+    oc = oc_map.get_cluster(site.cluster.name)
+    for item in oc.get_items(
+        kind="Secret",
+        namespace=site.namespace.name,
+        labels=site.token_labels,
+    ):
+        if item["metadata"]["name"] not in [
+            site.token_name(connected_site) for connected_site in site.connected_sites
+        ]:
```

Comment:
those token should be revoked on the remote site as well, right ?

Reply:
This removes all tokens from sites where the current site should not be connected anymore, e.g., the site doesn't exist anymore. Such kind of token doesn't exist on the source site anymore, because it has been delete during the token transfer. See [_transfer_token](https://github.com/app-sre/qontract-reconcile/blob/ae85b58f16cf564b3c0a1bef80e9f0550593d7d3/reconcile/skupper_network/reconciler.py#L115)

Reply:
but does it revoke the tokens as well ?

Reply:
I don't get it :( How do I revoke a connection-token? Especially if the source site doesn't exist anymore.

Reply:
i may be misunderstanding the term "source site". Any token deletion should go with a token revocation on the site that generated it. If that site is deleted, it's fine of course, but if the deletion occurs on a connecting site, we need a revoke somehow, right ?

Reply:
I don't think it's possible to revoke a single connection token. The `skupper` CLI has a `revoke-access` subcommand that revokes all previously issued tokens:

```
$ skupper revoke-access --help
This will invalidate all previously issued tokens and require that all
links to this site be re-established with new tokens.
```

The [code](https://github.com/skupperproject/skupper/blob/master/client/revoke_all.go#L124) behind that regenerates the site CA and all site certificates.